modules/tinymce/src/core/test/ts/browser/lookup/UserLookupTest.ts

```diff
@@ -0,0 +1,371 @@
+        name: 'Test User',
+        avatar: 'test-avatar.png',
+        description: 'Test Description',
+        custom: {
+          role: 'admin',
+          department: 'IT'
+        }
+      };
+
+      // Override fetch_users for this test
+      editor.options.set('fetch_users', () => Promise.resolve([ customUser ]));
+
+      const [ userPromise ] = editor.userLookup.fetchUsers([ userId ]);
+
+      await expect(userPromise).to.eventually.deep.equal(customUser);
+      await expect(userPromise).to.eventually.have.deep.property('custom', customUser.custom);
+    });
+
+    it('TINY-11974: Should handle network failures gracefully', async () => {
+      const editor = hook.editor();
+      const userId = 'test-user-network-error';
+
+      // Override fetch_users to simulate network failure
+      editor.options.set('fetch_users', () => Promise.reject(new Error('Network error')));
+
+      const [ userPromise ] = editor.userLookup.fetchUsers([ userId ]);
+
+      await expect(userPromise).to.be.rejectedWith('Network error');
+    });
+
+    it('TINY-11974: Should handle malformed server responses', async () => {
+      const editor = hook.editor();
+      const userId = 'test-user-malformed';
+
+      // Override fetch_users to return invalid data
+      editor.options.set('fetch_users', () => Promise.resolve([{
+        invalid: 'data'
+      }] as any));
+
+      const [ userPromise ] = editor.userLookup.fetchUsers([ userId ]);
+
+      await expect(userPromise).to.be.rejectedWith(`User ${userId} not found`);
+    });
+
+    it('TINY-11974: Should handle repeated requests for non-existent users', async () => {
+      const editor = hook.editor();
+      const nonExistentId = 'non-existent-user';
+      let fetchCount = 0;
+
+      // Clear any existing user cache
+      editor.userLookup = createUserLookup(editor);
+
+      // Override fetch_users to track the calls
+      editor.options.set('fetch_users', () => {
+        fetchCount++;
+        return Promise.resolve([]);
+      });
+
+      // Trigger fetch
+      const [ firstPromise ] = editor.userLookup.fetchUsers([ nonExistentId ]);
+
+      // Use cache
+      const [ secondPromise ] = editor.userLookup.fetchUsers([ nonExistentId ]);
+
+      await expect(firstPromise).to.eventually.be.rejectedWith(`User ${nonExistentId} not found`);
+      await expect(secondPromise).to.eventually.be.rejectedWith(`User ${nonExistentId} not found`);
+
+      // fetch_users should only have been called once during the initial fetch
+      // and not during the cache lookup for the second promise
+      expect(fetchCount).to.equal(1, 'Should only fetch once');
+    });
+
+    it('TINY-11974: Should maintain separate caches for different user IDs', async () => {
+      const editor = hook.editor();
+      let fetchCount = 0;
+
+      editor.userLookup = createUserLookup(editor);
+
+      editor.options.set('fetch_users', (userIds: string[]): Promise<User[]> => {
+        fetchCount++;
```

Comment:
You could use the TestStore here so track what userIds got hit a fetchCount could hide say bugs where the userIds passed into fetch isn't the expected ones in a refactor.